### ISSUE TITLE
Invalid PKCS/Bad Signature cause?

### ISSUE THREAD

**Issue by kevinh-csalabs** [Ylianst/MeshCentral#2197]
I am seeing "Invalid PKCS signature" and "Bad Signature" in the server info, but don't see any documentation on what would cause this or what needs to be done to resolve it. Any suggestions?

**Comment by vitko-bg** [comment]
Same here... After latest update to 0.7.48

**Comment by vitko-bg** [comment]
@kevinh-csalabs I am using currently this line in config.json to get agents connected and will try to update all agents, will see what will be the result
 "ignoreagenthashcheck": [ "86.134.X.X", "86.128.X.X", "86.8.X.X" ],

To get the IPs - run systemctl status meshcentral - it will show you which IPs are problematic.

**Comment by Ylianst** [comment]
By crazy chance, if you are using MongoDB, what version are you using? Also, try updating to v0.7.55 and let me know if it fixes it. I had MongoDB performance improvements starting in v0.7.46 and just pulled them all out in v0.7.55. I would love to know if it makes a difference.

As for documentation of this issue, I had never planned to see this error except if a hacker is purposefully trying to connect and inject junk into the websocket. In that case, the agent authentication signature would fail. This is certainly not something that should be happening.

**Comment by LPJon** [comment]
> 
> 
> By crazy chance, if you are using MongoDB, what version are you using? Also, try updating to v0.7.55 and let me know if it fixes it. I had MongoDB performance improvements starting in v0.7.46 and just pulled them all out in v0.7.55. I would love to know if it makes a difference.
> 
> As for documentation of this issue, I had never planned to see this error except if a hacker is purposefully trying to connect and inject junk into the websocket. In that case, the agent authentication signature would fail. This is certainly not something that should be happening.

@Ylianst  I agree. Man in the middle attacks to produce this as well as certain firewalls or compromised firewalls.

**Comment by Allowee** [comment]
Interesting.
I run everything on the same port, could it be port scanners or other tools trying to break in?

are there settings so I can log this and figure out what is happening?

**Comment by kevinh-csalabs** [comment]
I am running mongo, but still seeing entries after 0.7.55 upgrade:

Invalid PKCS signature: 9

Could this be due to running behind cloudflare?

**Comment by black3dynamite** [comment]
> I am running mongo, but still seeing entries after 0.7.55 upgrade:
> 
> Invalid PKCS signature: 9
> 
> Could this be due to running behind cloudflare?

It's possible. I was originally using Lets Encrypt with nginx (reverse proxy) but switched to cloudflare certs with nginx and started noticing the errors.

**Comment by kevinh-csalabs** [comment]
I’m using cloudflare but still using lets encrypt on the backend (confirmed cert are working and renewing as expected before turning on cloudflare)
________________________________

**Comment by si458** [comment]
@Smiley-k what message are you actually seeing? as the messages all mean different things.

**Comment by si458** [comment]
@Smiley-k nothing. Basically meshcentral is just telling u remote devices are trying to connect BUT something is wrong with the agents, so u need to uninstall the meshagent on the remote devices, redownload the agents and then reinstall

**Comment by si458** [comment]
@Smiley-k the only thing u COULD try (not 100% sure if it will work but case of trying it).
Set `ignoreagenthashcheck: true` to allow all ur devices to connect.
Then go into a device then the console tab and run `agentupdate`.
This should force download a copy of the agent and then restart it!
Then once that machine has restarted, try setting `ignoreagenthashcheck: false` and see if the device is online?

**Comment by si458** [comment]
@Smiley-k but it could also be a problem with ur reverse proxy/meshcentral setup if u use this, as the webserver ssl cert is different the meshagent ssl cert. So without knowing a little more about ur setup/config.json it's hard to diagnose the problem

**Comment by si458** [comment]
@Smiley-k so it could just be a config issue with nginx or ur meshcentral, again, plz share ur config.json and ur nginx config and th3 community can belp

**Comment by si458** [comment]
@Smiley-k

your `tlsoffload` should be the IP address of your reverse nginx server, not the website url

`agentidletimeout` shouldn't really be set as its default is 150 seconds anyway before it disconnects remote devices as they didn't send any web socket traffic

Only 2 issues I can find from a quick read, try those, restart meshcentral and see if it helps

**Comment by si458** [comment]
Try the following,

Backup then delete the `agents, signedagents` folders from inside `meshcentral-data`
Restart meshcentral and Make sure it says its resigning the exes.
then completely uninstall the agent from a device,
then redownload the agent from the web ui on the remote device,
Then reinstall 

It might be corrupt exes that have bee signed?

**Comment by Xevichr** [comment]
Currently having an "Invalid PKCS signature" and a "bad signature" along with it. Is there a way to tell which machine has these errors? 

I'm aware of `agentstats` and `agentissues`, but they don't help me determine which machine is having the errors. At least not with my knowledge. Maybe it is not possible in the first place?

I'm going to assume simply reinstalling the Agent will fix it once the machine having the issue is found.

**Comment by si458** [comment]
sadly the code only simply logs the remote ipaddress and port of the connection that had the `Invalid PKCS signature`
we would need to add a method into the code to store this extra info

**Comment by Xevichr** [comment]
When running `agentissues` I have the following result:
6/26/2024, 4:56:00 AM, , invalidPkcsSignature
6/26/2024, 4:56:00 AM, , BadSignature2
6/26/2024, 8:17:10 AM, , invalidPkcsSignature
6/26/2024, 8:17:10 AM, , BadSignature2

So this information is the only thing that will be send to MeshCentral, or is there a way I can find in MeshCentral, which IP had this issue? Because we are currently running MeshCentral LAN only, IP address would be more than enough to find the machine having the issue.

Besides that, considering I haven't even found the machines that have the issue yet, what are the consquences? I'm assuming simply no MeshAgent to MeshCentral connection?

**Comment by si458** [comment]
@Xevichr yes sorry, if you used the meshcentralinstaller, you need to STOP the meshcentral service in the windows services
THEN run that command, otherwise it runs on the next available port if the port it wants isnt available

**Comment by si458** [comment]
@Xevichr also you need to run `node node_modules/meshcentral --debug agent` dont change the word `agent`!
the `--debug` has multiple options to debug things like `--debug web,amt,agent`

**Comment by Xevichr** [comment]
Yeah, that both makes a lot of sense. In any case, thanks, I'll try it out and let you know if I find anything.

**Comment by jfiedlerNS** [comment]
Hey, was there any conclusion regarding this issue? We are seeing invalid PKCS signatures and bad signatures aswell. Running version 1.1.33, no proxy, Let's encrypt cert.

**Comment by jfiedlerNS** [comment]
So bit of an update - Linux server hosting meshcentral was pending restart, after it came back up no invalid signatures warnings as of yet.

**Comment by hide-afk** [comment]
> So bit of an update - Linux server hosting meshcentral was pending restart, after it came back up no invalid signatures warnings as of yet.

Hi @jfiedlerNS , I have Meshcentral 1.1.44, and still having both errors, do you have them too?

![Image](https://github.com/user-attachments/assets/911fdbc9-cbbf-4ca6-a399-b05b3d523b46)

**Comment by jfiedlerNS** [comment]
Hey @hide-afk ,
Unfortunately yes. We are running 1.1.44 as well and experiencing same issues. Turned out that after each reset of the server the error count resets to zero but it starts to climb after a while.

**Comment by hide-afk** [comment]
@jfiedlerNS @si458 
That's not very nice 😢 , It would be nice to investigate this issue and try to fix it or find a way to log which agent (with their MAC address or device name) is having the problem.
I am going to try these solutions soon:
https://github.com/Ylianst/MeshCentral/issues/2197#issuecomment-2067625665
https://github.com/Ylianst/MeshCentral/issues/2197#issuecomment-2066764670

**Comment by jfiedlerNS** [comment]
FIY I think I managed to track a few devices that were not responding, but after reinstalling agents on them they worked for a while and then stopped again.

**Comment by hide-afk** [comment]
> FIY I think I managed to track a few devices that were not responding, but after reinstalling agents on them they worked for a while and then stopped again.

I am having issues with agents that don't seem active, but they are. I guess the Invalid PKCS/Bad Signature errors are those agents not sending signals:

![Image](https://github.com/user-attachments/assets/802c112e-d8ac-4073-9c60-7b836bea3300)
![Image](https://github.com/user-attachments/assets/96efd63a-f642-4f0e-8143-f99073b050f0)

It's frustrating because I have tried to uninstall + install, update, etc. And still not working.

**Comment by jfiedlerNS** [comment]
Yeah, I came to the same conclussion. Some devices stop responding for no apparent reason, they were not even restarted. Restart of the device never helped, reinstall of the agent helped sometimes but the problem returned after a while at least for some of the devices. I get a lot of duplicateAgent errors when running agentissues command from console.

**Comment by si458** [comment]
@hide-afk if you enter `agentissues` it should help narrow down IP addresses and ports in firewalls
you can then check your firewall of that remote ip and see what device caused the issue, its a good starting point 👍 

also without knowing a little about your setup its hard to say
server os
mesh version
node version
where is mesh hosted?
what your config.json look like?
do you use a reverse proxy?
do you use a tunnel (ngrok, cloudlfare tunnels)?
where is the remote device located network wise? (on same as mesh or external)
what os is the remote device?
what mesh version is the remote device running?

**Comment by hide-afk** [comment]
hi @si458 , as the agents are in two offices and the `agentissues` command only shows public IP addresses, it doesn't help us that much. (we don't have a firewall installed in the offices)

some info about my installation:
Ubuntu Server 24.04 LTS
Hosted in AWS
No reverse proxy, no tunnel.
We have like 100 devices, 50% in one office, and 50% in the other.
All the devices are Windows 11.
I should check the mesh version in the agents that are showing failures, but I tried updating and reinstalling and still failing.

```
Server info:
{
    "meshVersion": "v1.1.44",
    "nodeVersion": "v18.19.1",
    "runMode": "WAN mode",
    "productionMode": true,
    "database": "MongoDB",
    "dbChangeStream": false,
    "dbBulkOperations": false,
    "platform": "linux",
    "arch": "x64",
    "pid": 11947,
    "uptime": 607779.894138982,
    "cpuUsage": {
        "user": 640837210,
        "system": 252904270
    },
    "memoryUsage": {
        "rss": 167927808,
        "heapTotal": 72257536,
        "heapUsed": 67395304,
        "external": 51062996,
        "arrayBuffers": 48796480
    },
    "warnings": [],
    "allDevGroupManagers": []
}
```

```
Config.json:
{
    "settings": {
        "MongoDb": "mongodb://127.0.0.1:27017/meshcentral",
        "WANonly": true,
        "port": 443,
        "redirPort": 80,
        "cert": "---",
        "sessionKey": "---",
        "log": ["*"],
        "logToFile": true,
        "logToMongo": true,
        "maxInvalidLogin": {
            "time": 15,
            "count": 4,
            "coolofftime": 15
        }
    },

    "domains": {
        "": {
            "title": "Remote",
            "titlePicture": "title.png",
            "welcomePicture": "welcome.png",
            "welcomeText": "Remote Desktop",
            "NewAccounts": false,
	    "showPasswordLogin": false,
            "authStrategies": {
                "google": {
                    "callbackurl": "---",
                    "clientid": "---",
                    "clientsecret": "---"
                }
            }
        }
    },

    "smtp": { 
        "host": "smtp.gmail.com",
        "port": 587,
        "from": "---.com",
        "user": "---.com",
        "pass": "---",
        "tls": false
    },
    
    "letsencrypt": {
        "email": "---.com",
        "names": "---.com",
        "production": true
    }
}
```

**Comment by si458** [comment]
@hide-afk so only thing ive spotted is you have 2 values in settings that dont exist in our source code anywhere but also dont think they would cause you issues
```
"logToFile": true, <- doesnt exist
"logToMongo": true, <- doesnt exist
```
also just discovered/learned that `log` can be an array (what you get for checking source code haha)
the schema says it can ONLY be a string hehe, so this needs fixing in the schema file 👍 

do the remote devices have to connect through a forward proxy (squid) at all to access the AWS?
also is the AWS an external ip or an internal ip? (VPC peering for example)

**Comment by hide-afk** [comment]
Oh I see @si458 , going to delete those lines in the next server reset, thank you! (at least it was useful for you to notice that small problem, hehe).

Regarding the questions, the remote devices aren't connecting through nothing special, they are configured as if they were normal devices (like personal use).
And the AWS instance is an external IP.

**Comment by jfiedlerNS** [comment]
I am missing these lines in the config so that really should not be an issue. We are also running it on ubuntu with no reverse proxy, but we have it self hosted, so shouldn't be AWS related either.
